security/advancedtls/advancedtls_test.go

```diff
@@ -679,3 +681,83 @@ func TestOptionsConfig(t *testing.T) {
 		})
 	}
 }
+
+func TestGetCertificateSNI(t *testing.T) {
+	// Load server certificates for setting the serverGetCert callback function.
+	serverPeerCert1, err := tls.LoadX509KeyPair(testdata.Path("server_cert_1.pem"),
+		testdata.Path("server_key_1.pem"))
+	if err != nil {
+		t.Fatalf("Server is unable to parse peer certificates. Error: %v", err)
+	}
+	serverPeerCert2, err := tls.LoadX509KeyPair(testdata.Path("server_cert_2.pem"),
+		testdata.Path("server_key_2.pem"))
+	if err != nil {
+		t.Fatalf("Server is unable to parse peer certificates. Error: %v", err)
+	}
+	serverPeerCert3, err := tls.LoadX509KeyPair(testdata.Path("server_cert_3.pem"),
+		testdata.Path("server_key_3.pem"))
+	if err != nil {
+		t.Fatalf("Server is unable to parse peer certificates. Error: %v", err)
+	}
+	tests := []struct {
+		desc                string
+		serverGetCert       func(*tls.ClientHelloInfo) ([]*tls.Certificate, error)
+		serverName          string
+		expectedCertificate tls.Certificate
+	}{
+		{
+			desc: "Selected certificate by SNI should be serverPeerCert1",
+			serverGetCert: func(info *tls.ClientHelloInfo) ([]*tls.Certificate, error) {
+				return []*tls.Certificate{&serverPeerCert1, &serverPeerCert2, &serverPeerCert3}, nil
+			},
+			// "foo.bar.com" is the common name on server certificate server_cert_1.pem.
+			serverName:          "foo.bar.com",
+			expectedCertificate: serverPeerCert1,
+		},
+		{
+			desc: "Selected certificate by SNI should be serverPeerCert2",
+			serverGetCert: func(info *tls.ClientHelloInfo) ([]*tls.Certificate, error) {
+				return []*tls.Certificate{&serverPeerCert1, &serverPeerCert2, &serverPeerCert3}, nil
+			},
+			// "foo.bar.server2.com" is the common name on server certificate server_cert_2.pem.
+			serverName:          "foo.bar.server2.com",
+			expectedCertificate: serverPeerCert2,
+		},
+		{
+			desc: "Selected certificate by SNI should be serverPeerCert3",
+			serverGetCert: func(info *tls.ClientHelloInfo) ([]*tls.Certificate, error) {
+				return []*tls.Certificate{&serverPeerCert1, &serverPeerCert2, &serverPeerCert3}, nil
+			},
+			// "google.com" is one of the DNS names on server certificate server_cert_3.pem.
+			serverName:          "google.com",
+			expectedCertificate: serverPeerCert3,
+		},
+	}
+	for _, test := range tests {
+		test := test
+		t.Run(test.desc, func(t *testing.T) {
+			serverOptions := &ServerOptions{
+				GetCertificate: test.serverGetCert,
+			}
+			serverConfig, err := serverOptions.config()
+			if err != nil {
+				t.Fatalf("Unable to generate serverConfig. Error: %v", err)
+			}
+			pointFormatUncompressed := uint8(0)
+			clientHello := &tls.ClientHelloInfo{
+				CipherSuites:      []uint16{tls.TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA},
```

Comment:
Are the other fields outside of `ServerName` useful for this test? If not, you can get rid of them.

Reply:
Yes, they are required to return the wanted result from `SupportsCertificate`. More details of the function logic can be found [here](https://github.com/golang/go/blob/d286e61b6787fe2b55bf0ec8a814962ebda8d202/src/crypto/tls/common.go#L1006-L1154) :)